tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/search/ModelCheckerParams.scala

```diff
@@ -43,8 +49,16 @@ class ModelCheckerParams(
    * The invariant checking mode. When it is equal to AfterJoin, the invariant is checked after joining all transitions
    * per step. When it is equal to BeforeJoin, the invariant is checked before joining all transitions.
    */
-  var invariantMode: InvariantMode =
-    if ("after" == tuningOptions.getOrElse("search.invariant.mode", "before")) AfterJoin else BeforeJoin
+  var invariantMode: ModelCheckerParams.InvariantMode = {
+    // we have experimentally found that the mode "atOnce" is more efficient for bounded model checking,
+    // whereas the mode "inLoop" is more efficient for invariant checking
+    val defaultMode = if (stepsBound <= 1) "inloop" else "atonce"
+    if (Set("after", "atonce").contains(tuningOptions.getOrElse("search.invariant.mode", defaultMode).toLowerCase)) {
+      ModelCheckerParams.CheckInvariantAtOnce()
+    } else {
+      ModelCheckerParams.CheckInvariantInLoop()
+    }
```

Comment:
Since the argument here is a string, should we also test in the else branch and throw on an invalid value?

Reply:
I am divided here. On one hand, you are right, we should indicate an error. On the other hand, I would prefer no to throw in case of an optimization option. Maybe log an warning?

Reply:
> On the other hand, I would prefer no to throw in case of an optimization option

Why not? Obviously the tool then does something other than the user intended.
We do know that users often miss warnings in the log ;)